Support Swift 3.0

norio-nomura opened the pull request:
- Support Swift 3.0 Preview 6
- Tests run on Xcode 8 beta 6 and Linux by using https://github.com/Quick/Quick/pull/510
- Tests does not run with Swift Package Manager on OS X.  
  Because https://github.com/Quick/Quick/pull/510 does not support that. I think we can't.
- Drop support Swift 2.x

norio-nomura commented:
I guessed that failing SPM test for OS X on Travis-CI is an issue of GitHub. https://bugs.swift.org/browse/SR-2070
But, it might be different issue…

norio-nomura commented:
The issue came from git-2.9.0 that fixed on git-2.9.1.
https://github.com/git/git/blob/master/Documentation/RelNotes/2.9.1.txt#L82
I updated to use a workaround.

ikesyo commented on Cartfile:
```diff
@@ -1 +1 @@
-github "antitypical/Result" ~> 2.1.1
+github "antitypical/Result" "master"
```
Can we use `3.0.0-alpha.1` now instead?

ikesyo commented on Commandant.xcodeproj/project.pbxproj:
```diff
@@ -430,6 +433,7 @@
 				PRODUCT_BUNDLE_IDENTIFIER = "org.carthage.$(PRODUCT_NAME:rfc1034identifier)";
 				PRODUCT_NAME = "$(TARGET_NAME)";
 				SWIFT_OPTIMIZATION_LEVEL = "-Onone";
+				SWIFT_VERSION = 3.0;
```
Setting the value at the project level, not at each target level, would be fine as same as #68.

ikesyo commented on Sources/Commandant/Errors.swift:
```diff
@@ -9,79 +9,79 @@
-#else
-	public typealias ClientErrorType = ErrorType
-#endif
+public typealias ClientErrorType = ErrorProtocol
```
Is the typealias not needed anymore (use `ErrorProtocol` directly)?

ikesyo commented on Sources/Commandant/ArgumentParser.swift:
```diff
@@ -96,15 +96,15 @@ public final class ArgumentParser {
 	/// If the key is found, it is then removed from the list of arguments
 	/// remaining to be parsed.
-	internal func consumeBooleanKey(key: String) -> Bool? {
+	internal func consumeBooleanKey(_ key: String) -> Bool? {
```
Maybe `consumeBoolean(key: String)`?

ikesyo commented on Sources/Commandant/ArgumentParser.swift:
```diff
@@ -120,7 +120,7 @@ public final class ArgumentParser {
 	/// If a value is found, the key and the value are both removed from the
 	/// list of arguments remaining to be parsed.
-	internal func consumeValueForKey(key: String) -> Result<String?, CommandantError<NoError>> {
+	internal func consumeValueForKey(_ key: String) -> Result<String?, CommandantError<NoError>> {
```
Maybe `consumeValue(forKey: String)`?

ikesyo commented on Sources/Commandant/ArgumentParser.swift:
```diff
@@ -162,25 +162,25 @@ public final class ArgumentParser {
 	/// Returns whether the given key was specified and removes it from the
 	/// list of arguments remaining.
-	internal func consumeKey(key: String) -> Bool {
+	internal func consumeKey(_ key: String) -> Bool {
```
Maybe `consume(key: String)`?

ikesyo commented on Sources/Commandant/ArgumentParser.swift:
```diff
@@ -162,25 +162,25 @@ public final class ArgumentParser {
 	/// Returns whether the given flag was specified and removes it from the
 	/// list of arguments remaining.
-	internal func consumeBooleanFlag(flag: Character) -> Bool {
+	internal func consumeBooleanFlag(_ flag: Character) -> Bool {
```
Maybe `consumeBoolean(flag: Character)`?

mdiep commented:
Looks good to me. Is this still a work in progress?

norio-nomura commented:
Swift 3.0 Preview 3 has already come. So I changed this to "Support Swift 3.0 Preview 3".
But, linux version is not yet tested on Swift 3.0 Preview 3.

And I also inspected to change for Swift 3 API Design Guidelines.
It would be:
- `ArgumentType` to `ArgumentProtocol`
  - `ArgumentType.fromString()` to `ArgumentType.from(string:)`
- `CommandType` to `CommandProtocol`
- `OptionsType` to `OptionsProtocol`

norio-nomura commented:
Sorry for confusing.
I reverted title to "Support Swift 3.0 Preview 2" and removed "[WIP]".

I will push another PR if Swift 3.0 Preview 3 will requires another changes on Linux, and also for Swift 3 API Design Guidelines

mdiep commented:
Oh, that's fine. We can wait and do the Swift 3 API changes here. I think it makes sense to do those all at once and there's no rush. But I don't think the linux changes need to be done here as that's still pretty alpha.

norio-nomura commented:
Oops! Sorry for missing each other. 😅

norio-nomura commented:
Thank you for giving smile :smile:.
I don't hurry about this. My main purpose of this PR is making upstream compatible with https://github.com/jpsim/SourceKitten/pull/223 on Linux.
So, I'll tagged this as "[WIP]" until I'll confirm on Swift 3.0 Preview 3 for Linux.

norio-nomura commented:
About "Swift 3 API Design Guidelines", I guess we need another PR for reviews.
Diffs to this branch is https://github.com/norio-nomura/Commandant/compare/nn-swift-3...norio-nomura:swift-3-api-design-guidelines

mdiep commented:
Thanks for tackling this! I don't think there's a rush for Carthage's sake, because Carthage will probably want to stay on Swift 2.3 for a bit. But it's great to have someone working on this. ✨

norio-nomura commented:
Update to Swift-3.0-Preview-6

norio-nomura commented:
I updated:
- Swift 3 API Design Guidelines
- Use `swift-3.0-RELEASE`

norio-nomura commented:
[The failed job](https://travis-ci.org/Carthage/Commandant/jobs/160100216) is caused by [SR-2651](https://bugs.swift.org/browse/SR-2651)

ikesyo commented:
@norio-nomura Could you apply the review https://github.com/Carthage/Commandant/pull/73#r78555764 to this pull request?

ikesyo commented on Package.swift:
```diff
@@ -1,8 +1,18 @@
+        #else
+            return [
+                .Package(url: "https://github.com/antitypical/Result.git", majorVersion: 3, minor: 0),
+                .Package(url: "https://github.com/ikesyo/Nimble", "5.0.0-alpha.30.gm.candidate"),
```
`v5.0.0-alpha.30` on ikesyo/Nimble can be used now. 🙏

norio-nomura commented:
updated.

norio-nomura commented:
Changes of Swift 3 API Design Guidelines:
- `ArgumentType` to `ArgumentProtocol`
  - `fromString()` to `from(string:)`
- `CommandType` to `CommandProtocol`
- `OptionsType` to `OptionsProtocol`
- `CommandRegistry`
  - `runCommand(_:arguments:)` to `run(command:arguments:)`

jpsim commented:
🎉

I suppose a release should be cut soon?

mdiep commented:
It can be if that'd be helpful! Carthage won't update for a while, since we'd like to keep 2.2 compatibility for a while. But if you want Swift 3 support, I'll create a release.

jpsim commented:
Yes, that would be appreciated! That'll mean I can move [SourceKitten's Linux support](https://github.com/jpsim/SourceKitten/pull/223) off my personal fork of Commandant!

mdiep commented:
https://github.com/Carthage/Commandant/releases/tag/0.11

jpsim commented:
Thanks!
